Cache npm packages in build bots

CI runs such as [this one](https://github.com/unicode-org/icu4x/pull/920/checks?check_run_id=3660897876) report the following errors:

```
Error: Error: TAR_BAD_ARCHIVE: Unrecognized archive format
    at main (/usr/local/lib/node_modules/wasm-opt/bin/index.js:119:11)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! wasm-opt@1.2.1 postinstall: `node bin/index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the wasm-opt@1.2.1 postinstall script.
```

This is likely due to npm being flaky when downloading dependencies.  We should create a package-lock.json with our required npm dependencies, and then add a caching step in GitHub Actions like we do with other dependencies.

--- a/.github/workflows/build-test.yml
+++ b/.github/workflows/build-test.yml
@@ -237,7 +237,16 @@ jobs:
           rustup target add wasm32-unknown-unknown --toolchain nightly-2021-12-22
           rustup target add wasm32-unknown-emscripten --toolchain nightly-2021-12-22
           rustup component add rust-src --toolchain nightly-2021-12-22
-      - name: Install WASM tools
+
+      - name: Install npm dependencies from cache
+        uses: actions/cache@v2
+        id: npm-cache
+        with:
+          path: ffi/diplomat/wasm/node_modules
+          key: npm-${{ hashFiles('ffi/diplomat/wasm/package-lock.json') }}
+      - name: Install npm packages.
+        # If npm dependencies are not found in cache.
+        if: steps.npm-cache.outputs.cache-hit != 'true'
         run: |
           # This step requires wasm-opt and wabt. To install these two packages
           # ready-to-use locally, you also can execute
@@ -245,8 +254,11 @@ jobs:
           # 'sudo npm install -g wabt'.
           cd ffi/diplomat/wasm
           npm ci
-          echo "$HOME/work/icu4x/icu4x/ffi/diplomat/wasm/node_modules/wasm-opt/bin" >> $GITHUB_PATH
-          echo "$HOME/work/icu4x/icu4x/ffi/diplomat/wasm/node_modules/wabt/bin" >> $GITHUB_PATH
+      - name: Set path to wasm-opt and wabt binaries
+        run: |
+          echo "$PWD/ffi/diplomat/wasm/node_modules/wasm-opt/bin" >> $GITHUB_PATH
+          echo "$PWD/ffi/diplomat/wasm/node_modules/wabt/bin" >> $GITHUB_PATH
+
       - name: Get twiggy version
         id: twiggy-version
         run: |
@@ -765,15 +777,34 @@ jobs:
         command: install
         args: --git https://github.com/rustwasm/twiggy.git --rev ${{ steps.twiggy-version.outputs.hash }} twiggy
 
+    - name: Install npm dependencies from cache
+      uses: actions/cache@v2
+      id: npm-cache
+      with:
+        path: ffi/diplomat/wasm/node_modules
+        key: npm-${{ hashFiles('ffi/diplomat/wasm/package-lock.json') }}
+
+    - name: Install npm packages.
+      # If npm dependencies are not found in cache.
+      if: steps.npm-cache.outputs.cache-hit != 'true'
+      # Note: if package.json and package-lock.json are out of sync then
+      # 'npm ci' will fail. Locally execute 'npm install' in directory
+      # 'ffi/diplomat/wasm' to generate a new package-lock.json file and
+      # merge it into the repository.
+      run: |
+        cd ffi/diplomat/wasm
+        npm ci
+
+    - name: Set path to wasm-opt and wabt binaries
+      run: |
+        echo "$PWD/ffi/diplomat/wasm/node_modules/wasm-opt/bin" >> $GITHUB_PATH
+        echo "$PWD/ffi/diplomat/wasm/node_modules/wabt/bin" >> $GITHUB_PATH
+
     - name: Install prerequisites for wasm build
       run: |
         rustup component add rust-src
         rustup toolchain list
         rustup toolchain install nightly-2021-12-22
-        cd ffi/diplomat/wasm
-        npm ci
-        echo "$HOME/work/icu4x/icu4x/ffi/diplomat/wasm/node_modules/wasm-opt/bin" >> $GITHUB_PATH
-        echo "$HOME/work/icu4x/icu4x/ffi/diplomat/wasm/node_modules/wabt/bin" >> $GITHUB_PATH
 
     - name: Setup output data directory
       run: |

--- a/.github/workflows/build-test.yml
+++ b/.github/workflows/build-test.yml
@@ -243,7 +243,7 @@ jobs:
         id: npm-cache
         with:
           path: ffi/diplomat/wasm/node_modules
-          key: npm-${{ hashFiles('ffi/diplomat/wasm/package-lock.json') }}
+          key: ${{ runner.os }}-npm-${{ hashFiles('ffi/diplomat/wasm/package-lock.json') }}
       - name: Install npm packages.
         # If npm dependencies are not found in cache.
         if: steps.npm-cache.outputs.cache-hit != 'true'
@@ -782,7 +782,7 @@ jobs:
       id: npm-cache
       with:
         path: ffi/diplomat/wasm/node_modules
-        key: npm-${{ hashFiles('ffi/diplomat/wasm/package-lock.json') }}
+        key: ${{ runner.os }}-npm-${{ hashFiles('ffi/diplomat/wasm/package-lock.json') }}
 
     - name: Install npm packages.
       # If npm dependencies are not found in cache.
@@ -794,6 +794,14 @@ jobs:
       run: |
         cd ffi/diplomat/wasm
         npm ci
+        if [ $? -ne 0 ]
+        then
+          echo Execution of npm ci failed. This may be due to package.json
+          echo and package-lock,json being out of sync. Locally execute
+          echo npm install in directory  ffi/diplomat/wasm to generate a
+          echo new package-lock.json file and and merge it into the repository.
+          exit 1
+        fi
 
     - name: Set path to wasm-opt and wabt binaries
       run: |

--- a/.github/workflows/build-test.yml
+++ b/.github/workflows/build-test.yml
@@ -797,7 +797,7 @@ jobs:
         if [ $? -ne 0 ]
         then
           echo Execution of npm ci failed. This may be due to package.json
-          echo and package-lock,json being out of sync. Locally execute
+          echo and package-lock.json being out of sync. Locally execute
           echo npm install in directory  ffi/diplomat/wasm to generate a
           echo new package-lock.json file and and merge it into the repository.
           exit 1

